Fix numpy warning when importing torch without numpy installed

mikaylagawarecki opened the pull request:
Fixes #115638

I verified locally that with no numpy install the warning no longer occurs

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115867




cc @albanD @mruberry @jbschlosser @walterddr

mikaylagawarecki committed:
Fix numpy warning when importing torch without numpy installed

[ghstack-poisoned]

soulitzer commented on torch/nn/modules/transformer.py:
```diff
@@ -17,13 +17,17 @@
     The masked positions are filled with float('-inf'). Unmasked positions are filled with float(0.0).
     """
+    if device is None:
+        device = torch.device('cpu')
```
Why aren't we using torch._C._get_default_device() anymore? Couldn't that also return cuda?

mikaylagawarecki replied:
@soulitzer sorry I was slow in commenting, see https://github.com/pytorch/pytorch/pull/115867#discussion_r1427290212

soulitzer replied:
Oh what I mean is to do something like:
```
def _generate_square_subsequent_mask(
        sz: int,
        device: Optional[torch.device] = None,
        dtype: Optional[torch.dtype] = None,
) -> Tensor:
    if device is None:
        device = torch._C._get_default_device()
   [...]
```
Does that work?

mikaylagawarecki replied:
My bad, I had thought that  `torch.set_default_device('cuda')` would change the type returned by `torch._C._get_default_device()` so I thought that the device of the tensor returned by `_generate_square_subsequent_mask` didn't change regardless of what I was setting the device to.

If I understand correctly seems that the only way to set the device gotten by `torch._C._get_default_device` is via `torch.set_default_tensor_type(torch.cuda.FloatTensor)` which is deprecated as of 2.1  😛  

Nevertheless changed these to still call the functions

soulitzer replied:
If only there were a way to get the status of whatever `torch.set_default_device` or `with torch.device("cuda")` sets though 🤔  
I'm also not sure what is `torch.set_default_tensor_type(torch.cuda.FloatTensor)` is supposed to do...
```
torch.set_default_tensor_type(torch.cuda.FloatTensor)
 print(torch._C._get_default_device())  # cuda
 a = torch.zeros(1)   # cpu
 ```

mikaylagawarecki commented on torch/nn/modules/transformer.py:
```diff
@@ -17,13 +17,17 @@
 
 def _generate_square_subsequent_mask(
         sz: int,
-        device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),
```
These functions should not be called as the default (and it doesn't make sense either to have them here because after importing changing the default device/dtype does not change it in the function

```
>>> import torch
>>> from torch.nn.modules.transformer import _generate_square_subsequent_mask
>>> _generate_square_subsequent_mask(2)
tensor([[0., -inf],
        [0., 0.]])
>>> torch.set_default_device('cuda')
# still on CPU
>>> _generate_square_subsequent_mask(2)
tensor([[0., -inf],
        [0., 0.]])
```

Changing the default to `torch.device('cpu') also seems to initialize numpy so I think setting it to `None` fixes the error in the most BC-compatible way

mikaylagawarecki commented on torch/nn/modules/transformer.py:
```diff
@@ -214,8 +218,8 @@ def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, t
             sz: int,
-            device: torch.device = torch.device(torch._C._get_default_device()),  # torch.device('cpu'),
-            dtype: torch.dtype = torch.get_default_dtype(),
+            device: Optional[torch.device] = None,
```
This always calls into `_generate_square_subsequent_mask` so we don't need to handle the `None`

mikaylagawarecki committed:
Update on "Fix numpy warning when importing torch without numpy installed"

Fix #115638

I verified locally that with no numpy install the warning no longer occurs




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix numpy warning when importing torch without numpy installed"

Fix #115638

I verified locally that with no numpy install the warning no longer occurs




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix numpy warning when importing torch without numpy installed"

Fix #115638

I verified locally that with no numpy install the warning no longer occurs




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Fix numpy warning when importing torch without numpy installed"

Fix #115638

I verified locally that with no numpy install the warning no longer occurs




cc gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen

[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge
